File: plugins/llm/run-claude/repos/go-litellm/internal/anthropic/translate.go
// Package anthropic translates Anthropic Messages ↔ OpenAI Chat for /v1/messages.
package anthropic

import (
	"encoding/json"
	"strings"

	"github.com/noizu-labs/go-litellm/internal/jsonx"
	"github.com/noizu-labs/go-litellm/internal/providers"
)

// RequestToOpenAI translates an Anthropic Messages body into an OpenAI chat body.
func RequestToOpenAI(body map[string]any) map[string]any {
	msgs := append(systemMessages(body["system"]), translateMessages(body["messages"])...)
	out := map[string]any{
		"model":    body["model"],
		"messages": msgs,
	}
	copyIf(out, body, "max_tokens", "")
	copyIf(out, body, "temperature", "")
	copyIf(out, body, "top_p", "")
	copyIf(out, body, "stream", "")
	copyIf(out, body, "stop_sequences", "stop")
	putTools(out, body["tools"])
	putToolChoice(out, jsonx.AsMap(body["tool_choice"]))
	putReasoningEffort(out, body)
	return jsonx.DropNil(out)
}

func putReasoningEffort(out, body map[string]any) {
	thinking := jsonx.AsMap(body["thinking"])
	if thinking == nil {
		return
	}
	if _, exists := out["reasoning_effort"]; exists {
		return
	}
	switch jsonx.Str(thinking, "type") {
	case "enabled", "adaptive":
		out["reasoning_effort"] = "high"
	case "disabled":
		out["reasoning_effort"] = "low"
	}
}

func systemMessages(system any) []any {
	switch s := system.(type) {
	case nil:
		return nil
	case string:
		return []any{map[string]any{"role": "system", "content": s}}
	case []any:
		var parts []string
		for _, b := range s {
			if m := jsonx.AsMap(b); m != nil && jsonx.Str(m, "type") == "text" {
				parts = append(parts, jsonx.Str(m, "text"))
			}
		}
		return []any{map[string]any{"role": "system", "content": strings.Join(parts, "\n")}}
	default:
		return nil
	}
}

func translateMessages(raw any) []any {
	list, _ := raw.([]any)
	var out []any
	for _, item := range list {
		out = append(out, translateMessage(jsonx.AsMap(item))...)
	}
	return out
}

func translateMessage(m map[string]any) []any {
	if m == nil {
		return nil
	}
	role := jsonx.Str(m, "role")
	switch c := m["content"].(type) {
	case string:
		return []any{map[string]any{"role": role, "content": c}}
	case []any:
		if role == "assistant" {
			return []any{assistantBlocks(c)}
		}
		if role == "user" {
			return userBlocks(c)
		}
	}
	return []any{m}
}

func assistantBlocks(blocks []any) map[string]any {
	var text strings.Builder
	var toolCalls []any
	for _, b := range blocks {
		m := jsonx.AsMap(b)
		if m == nil {
			continue
		}
		switch jsonx.Str(m, "type") {
		case "text":
			text.WriteString(jsonx.Str(m, "text"))
		case "tool_use":
			args, _ := json.Marshal(m["input"])
			if string(args) == "null" {
				args = []byte("{}")
			}
			toolCalls = append(toolCalls, map[string]any{
				"id":   m["id"],
				"type": "function",
				"function": map[string]any{
					"name":      m["name"],
					"arguments": string(args),
				},
			})
		}
	}
	msg := map[string]any{"role": "assistant"}
	if t := text.String(); t != "" {
		msg["content"] = t
	} else {
		msg["content"] = nil
	}
	if len(toolCalls) > 0 {
		msg["tool_calls"] = toolCalls
	}
	return msg
}

func userBlocks(blocks []any) []any {
	var toolMsgs []any
	var rest []any
	for _, b := range blocks {
		m := jsonx.AsMap(b)
		if m != nil && jsonx.Str(m, "type") == "tool_result" {
			toolMsgs = append(toolMsgs, map[string]any{
				"role":         "tool",
				"tool_call_id": m["tool_use_id"],
				"content":      toolResultContent(m["content"]),
			})
			continue
		}
		rest = append(rest, b)
	}
	userContent := userContentOf(rest)
	var out []any
	out = append(out, toolMsgs...)
	if userContent != nil && userContent != "" {
		if list, ok := userContent.([]any); ok && len(list) == 0 {
			return out
		}
		out = append(out, map[string]any{"role": "user", "content": userContent})
	}
	return out
}

func toolResultContent(c any) string {
	switch t := c.(type) {
	case string:
		return t
	case []any:
		var parts []string
		for _, b := range t {
			if m := jsonx.AsMap(b); m != nil && jsonx.Str(m, "type") == "text" {
				parts = append(parts, jsonx.Str(m, "text"))
			} else {
				raw, _ := json.Marshal(b)
				parts = append(parts, string(raw))
			}
		}
		return strings.Join(parts, "\n")
	default:
		raw, _ := json.Marshal(c)
		return string(raw)
	}
}

func userContentOf(blocks []any) any {
	allText := true
	for _, b := range blocks {
		m := jsonx.AsMap(b)
		if m == nil || jsonx.Str(m, "type") != "text" {
			allText = false
			break
		}
	}
	if allText {
		var s strings.Builder
		for _, b := range blocks {
			s.WriteString(jsonx.Str(jsonx.AsMap(b), "text"))
		}
		return s.String()
	}
	var parts []any
	for _, b := range blocks {
		m := jsonx.AsMap(b)
		if m == nil {
			continue
		}
		switch jsonx.Str(m, "type") {
		case "text":
			parts = append(parts, map[string]any{"type": "text", "text": m["text"]})
		case "image":
			src := jsonx.AsMap(m["source"])
			if jsonx.Str(src, "type") == "base64" {
				url := "data:" + jsonx.Str(src, "media_type") + ";base64," + jsonx.Str(src, "data")
				parts = append(parts, map[string]any{"type": "image_url", "image_url": map[string]any{"url": url}})
			} else {
				parts = append(parts, map[string]any{"type": "image_url", "image_url": map[string]any{"url": jsonx.Str(src, "url")}})
			}
		default:
			raw, _ := json.Marshal(m)
			parts = append(parts, map[string]any{"type": "text", "text": string(raw)})
		}
	}
	return parts
}

func putTools(body map[string]any, tools any) {
	list, ok := tools.([]any)
	if !ok {
		return
	}
	out := make([]any, 0, len(list))
	for _, t := range list {
		m := jsonx.AsMap(t)
		schema := m["input_schema"]
		if schema == nil {
			schema = map[string]any{"type": "object", "properties": map[string]any{}}
		}
		out = append(out, map[string]any{
			"type": "function",
			"function": map[string]any{
				"name":        m["name"],
				"description": m["description"],
				"parameters":  schema,
			},
		})
	}
	body["tools"] = out
}

func putToolChoice(body map[string]any, tc map[string]any) {
	if tc == nil {
		return
	}
	switch jsonx.Str(tc, "type") {
	case "auto":
		body["tool_choice"] = "auto"
	case "any":
		body["tool_choice"] = "required"
	case "tool":
		body["tool_choice"] = map[string]any{
			"type":     "function",
			"function": map[string]any{"name": tc["name"]},
		}
	}
}

func copyIf(acc, body map[string]any, key, as string) {
	if v, ok := body[key]; ok && v != nil {
		if as == "" {
			as = key
		}
		acc[as] = v
	}
}

// ResponseFromModelResponse translates OpenAI chat → Anthropic Messages.
func ResponseFromModelResponse(r *providers.ModelResponse, requestedModel string) map[string]any {
	var choice map[string]any
	if r != nil && len(r.Choices) > 0 {
		choice = jsonx.AsMap(r.Choices[0])
	}
	if choice == nil {
		choice = map[string]any{}
	}
	message := jsonx.AsMap(choice["message"])
	if message == nil {
		message = map[string]any{}
	}
	id := ""
	if r != nil {
		id = r.ID
	}
	if id == "" {
		id = providers.RandID(12)
	}
	return map[string]any{
		"id":            "msg_" + id,
		"type":          "message",
		"role":          "assistant",
		"model":         requestedModel,
		"content":       buildContentBlocks(message),
		"stop_reason":   StopReason(strField(choice, "finish_reason")),
		"stop_sequence": nil,
		"usage": Usage(jsonx.AsMap(func() any {
			if r == nil {
				return nil
			}
			return r.Usage
		}())),
	}
}

func buildContentBlocks(message map[string]any) []any {
	var blocks []any
	reasoning := jsonx.Str(message, "reasoning")
	if reasoning == "" {
		reasoning = jsonx.Str(message, "reasoning_content")
	}
	if reasoning != "" {
		blocks = append(blocks, map[string]any{"type": "thinking", "thinking": reasoning})
	}
	text := jsonx.Str(message, "content")
	if text != "" {
		blocks = append(blocks, map[string]any{"type": "text", "text": text})
	}
	var toolCalls []any
	switch t := message["tool_calls"].(type) {
	case []any:
		toolCalls = t
	}
	for _, tc := range toolCalls {
		m := jsonx.AsMap(tc)
		fun := jsonx.AsMap(m["function"])
		id := jsonx.Str(m, "id")
		if id == "" {
			id = "toolu_" + providers.RandID(12)
		}
		blocks = append(blocks, map[string]any{
			"type":  "tool_use",
			"id":    id,
			"name":  jsonx.Str(fun, "name"),
			"input": decodeArgs(fun["arguments"]),
		})
	}
	if len(blocks) == 0 {
		return []any{map[string]any{"type": "text", "text": ""}}
	}
	return blocks
}

func decodeArgs(args any) any {
	switch t := args.(type) {
	case nil:
		return map[string]any{}
	case map[string]any:
		return t
	case string:
		var m map[string]any
		if err := json.Unmarshal([]byte(t), &m); err == nil {
			return m
		}
		return map[string]any{}
	default:
		return map[string]any{}
	}
}

func strField(m map[string]any, k string) string {
	if s, ok := m[k].(string); ok {
		return s
	}
	return ""
}

// StopReason maps an OpenAI finish_reason to Anthropic stop_reason.
func StopReason(r string) string {
	switch r {
	case "stop":
		return "end_turn"
	case "length":
		return "max_tokens"
	case "tool_calls":
		return "tool_use"
	case "content_filter":
		return "refusal"
	case "":
		return "end_turn"
	default:
		return r
	}
}

// Usage maps OpenAI usage → Anthropic usage.
func Usage(u map[string]any) map[string]any {
	if u == nil {
		return map[string]any{"input_tokens": 0, "output_tokens": 0}
	}
	in, _ := jsonx.Int(u, "prompt_tokens")
	out, _ := jsonx.Int(u, "completion_tokens")
	return map[string]any{"input_tokens": in, "output_tokens": out}
}

// StreamPreamble is message_start + content_block_start.
func StreamPreamble(model string) []string {
	message := map[string]any{
		"id": "msg_" + providers.RandID(12), "type": "message", "role": "assistant",
		"model": model, "content": []any{}, "stop_reason": nil, "stop_sequence": nil,
		"usage": map[string]any{"input_tokens": 0, "output_tokens": 0},
	}
	return []string{
		SSE("message_start", map[string]any{"type": "message_start", "message": message}),
		SSE("content_block_start", map[string]any{
			"type": "content_block_start", "index": 0,
			"content_block": map[string]any{"type": "text", "text": ""},
		}),
	}
}

// StreamTextDelta is one content_block_delta.
func StreamTextDelta(text string) string {
	return SSE("content_block_delta", map[string]any{
		"type": "content_block_delta", "index": 0,
		"delta": map[string]any{"type": "text_delta", "text": text},
	})
}

// StreamClosing is content_block_stop + message_delta + message_stop.
func StreamClosing(finishReason string, usage map[string]any) []string {
	outTok, _ := jsonx.Int(usage, "completion_tokens")
	return []string{
		SSE("content_block_stop", map[string]any{"type": "content_block_stop", "index": 0}),
		SSE("message_delta", map[string]any{
			"type":  "message_delta",
			"delta": map[string]any{"stop_reason": StopReason(finishReason), "stop_sequence": nil},
			"usage": map[string]any{"output_tokens": outTok},
		}),
		SSE("message_stop", map[string]any{"type": "message_stop"}),
	}
}

// SSE formats one Anthropic SSE frame.
func SSE(event string, data any) string {
	b, _ := json.Marshal(data)
	return "event: " + event + "\ndata: " + string(b) + "\n\n"
}

File: plugins/llm/run-claude/repos/ex-litellm/lib/ex_litellm/anthropic/translate.ex
defmodule ExLiteLLM.Anthropic.Translate do
  @moduledoc """
  Bidirectional Anthropic Messages ↔ OpenAI Chat translation for the
  `/v1/messages` endpoint — litellm's `anthropic_endpoints` adapter.

  Claude Code (and any Anthropic-SDK client) speaks the Messages API. When the
  requested model resolves to an OpenAI-family provider, the gateway translates:

    * **request** — Anthropic Messages body → OpenAI chat body (`system` field →
      system message, content blocks → OpenAI content/tool messages, `tools`
      input_schema → parameters, `stop_sequences` → `stop`).
    * **response** — OpenAI `ModelResponse` → Anthropic message (content blocks,
      `tool_calls` → `tool_use`, `finish_reason` → `stop_reason`, and crucially
      `usage.prompt_tokens/completion_tokens` → `usage.input_tokens/output_tokens`
      — Anthropic clients hard-depend on those fields).
    * **stream** — normalized provider chunks → Anthropic SSE events
      (`message_start` → `content_block_*` → `message_delta` → `message_stop`).
  """

  alias ExLiteLLM.Core.ModelResponse

  # === request: Anthropic → OpenAI ===

  @doc "Translate an Anthropic Messages request body into an OpenAI chat body."
  @spec request_to_openai(map()) :: map()
  def request_to_openai(body) do
    messages = system_messages(body["system"]) ++ translate_messages(body["messages"] || [])

    %{"model" => body["model"], "messages" => messages}
    |> copy(body, "max_tokens")
    |> copy(body, "temperature")
    |> copy(body, "top_p")
    |> copy(body, "stream")
    |> copy(body, "stop_sequences", "stop")
    |> put_tools(body["tools"])
    |> put_tool_choice(body["tool_choice"])
    |> put_reasoning_effort(body)
    |> Map.reject(fn {_k, v} -> is_nil(v) end)
  end

  defp system_messages(nil), do: []
  defp system_messages(system) when is_binary(system), do: [%{"role" => "system", "content" => system}]

  defp system_messages(blocks) when is_list(blocks) do
    text = Enum.map_join(blocks, "\n", fn
      %{"type" => "text", "text" => t} -> t
      _ -> ""
    end)

    [%{"role" => "system", "content" => text}]
  end

  defp translate_messages(messages) do
    Enum.flat_map(messages, &translate_message/1)
  end

  # A single Anthropic message can expand to several OpenAI messages (e.g. an
  # assistant turn with text + tool_use, or a user turn carrying tool_results).
  defp translate_message(%{"role" => role, "content" => content}) when is_binary(content) do
    [%{"role" => role, "content" => content}]
  end

  defp translate_message(%{"role" => "assistant", "content" => blocks}) when is_list(blocks) do
    text =
      blocks
      |> Enum.filter(&(&1["type"] == "text"))
      |> Enum.map_join("", & &1["text"])

    tool_calls =
      blocks
      |> Enum.filter(&(&1["type"] == "tool_use"))
      |> Enum.map(fn b ->
        %{
          "id" => b["id"],
          "type" => "function",
          "function" => %{"name" => b["name"], "arguments" => Jason.encode!(b["input"] || %{})}
        }
      end)

    msg = %{"role" => "assistant", "content" => (if text == "", do: nil, else: text)}
    msg = if tool_calls == [], do: msg, else: Map.put(msg, "tool_calls", tool_calls)
    [msg]
  end

  defp translate_message(%{"role" => "user", "content" => blocks}) when is_list(blocks) do
    {tool_results, rest} = Enum.split_with(blocks, &(&1["type"] == "tool_result"))

    tool_msgs =
      Enum.map(tool_results, fn tr ->
        %{
          "role" => "tool",
          "tool_call_id" => tr["tool_use_id"],
          "content" => tool_result_content(tr["content"])
        }
      end)

    user_content = user_blocks(rest)
    user_msgs = if user_content in [nil, "", []], do: [], else: [%{"role" => "user", "content" => user_content}]

    tool_msgs ++ user_msgs
  end

  defp translate_message(other), do: [other]

  defp tool_result_content(content) when is_binary(content), do: content

  defp tool_result_content(blocks) when is_list(blocks) do
    Enum.map_join(blocks, "\n", fn
      %{"type" => "text", "text" => t} -> t
      other -> Jason.encode!(other)
    end)
  end

  defp tool_result_content(other), do: Jason.encode!(other)

  # User content blocks → OpenAI content (string when all-text, else parts).
  defp user_blocks(blocks) do
    if Enum.all?(blocks, &(&1["type"] == "text")) do
      Enum.map_join(blocks, "", & &1["text"])
    else
      Enum.map(blocks, fn
        %{"type" => "text", "text" => t} ->
          %{"type" => "text", "text" => t}

        %{"type" => "image", "source" => %{"type" => "base64", "media_type" => mt, "data" => data}} ->
          %{"type" => "image_url", "image_url" => %{"url" => "data:#{mt};base64,#{data}"}}

        %{"type" => "image", "source" => %{"type" => "url", "url" => url}} ->
          %{"type" => "image_url", "image_url" => %{"url" => url}}

        other ->
          %{"type" => "text", "text" => Jason.encode!(other)}
      end)
    end
  end

  defp put_tools(body, nil), do: body

  defp put_tools(body, tools) when is_list(tools) do
    Map.put(
      body,
      "tools",
      Enum.map(tools, fn t ->
        %{
          "type" => "function",
          "function" => %{
            "name" => t["name"],
            "description" => t["description"],
            "parameters" => t["input_schema"] || %{"type" => "object", "properties" => %{}}
          }
        }
      end)
    )
  end

  defp put_tool_choice(body, nil), do: body
  defp put_tool_choice(body, %{"type" => "auto"}), do: Map.put(body, "tool_choice", "auto")
  defp put_tool_choice(body, %{"type" => "any"}), do: Map.put(body, "tool_choice", "required")

  defp put_tool_choice(body, %{"type" => "tool", "name" => name}),
    do: Map.put(body, "tool_choice", %{"type" => "function", "function" => %{"name" => name}})

  defp put_tool_choice(body, _), do: body

  # Anthropic `thinking` is not an OpenAI field. Map it onto reasoning_effort
  # so gpt-oss / similar models actually spend CoT budget. Deployment YAML
  # overwrites this later (haiku=low, opus=high).
  defp put_reasoning_effort(acc, %{"thinking" => %{"type" => t}}) when t in ["enabled", "adaptive"] do
    Map.put_new(acc, "reasoning_effort", "high")
  end

  defp put_reasoning_effort(acc, %{"thinking" => %{"type" => "disabled"}}) do
    Map.put_new(acc, "reasoning_effort", "low")
  end

  defp put_reasoning_effort(acc, _), do: acc

  defp copy(acc, body, key, as \\ nil) do
    case Map.get(body, key) do
      nil -> acc
      v -> Map.put(acc, as || key, v)
    end
  end

  # === response: ModelResponse → Anthropic message ===

  @doc "Translate a normalized OpenAI ModelResponse into an Anthropic Messages response."
  @spec response_from_model_response(ModelResponse.t(), String.t()) :: map()
  def response_from_model_response(%ModelResponse{} = r, requested_model) do
    choice = List.first(r.choices) || %{}
    message = choice_field(choice, :message) || %{}
    content_blocks = build_content_blocks(message)

    %{
      "id" => "msg_" <> (r.id || rand()),
      "type" => "message",
      "role" => "assistant",
      "model" => requested_model,
      "content" => content_blocks,
      "stop_reason" => stop_reason(choice_field(choice, :finish_reason)),
      "stop_sequence" => nil,
      "usage" => usage(r.usage)
    }
  end

  defp build_content_blocks(message) do
    text = message["content"] || message[:content]
    reasoning = message["reasoning"] || message[:reasoning] || message["reasoning_content"] || message[:reasoning_content]
    tool_calls = message["tool_calls"] || message[:tool_calls] || []

    thinking_blocks =
      if reasoning in [nil, ""], do: [], else: [%{"type" => "thinking", "thinking" => reasoning}]

    text_blocks = if text in [nil, ""], do: [], else: [%{"type" => "text", "text" => text}]

    tool_blocks =
      Enum.map(tool_calls, fn tc ->
        fun = tc["function"] || tc[:function] || %{}

        %{
          "type" => "tool_use",
          "id" => tc["id"] || tc[:id] || "toolu_" <> rand(),
          "name" => fun["name"] || fun[:name],
          "input" => decode_arguments(fun["arguments"] || fun[:arguments])
        }
      end)

    case thinking_blocks ++ text_blocks ++ tool_blocks do
      [] -> [%{"type" => "text", "text" => ""}]
      blocks -> blocks
    end
  end

  defp decode_arguments(nil), do: %{}

  defp decode_arguments(args) when is_binary(args) do
    case Jason.decode(args) do
      {:ok, map} when is_map(map) -> map
      _ -> %{}
    end
  end

  defp decode_arguments(args) when is_map(args), do: args

  @doc "Map an OpenAI finish_reason to an Anthropic stop_reason."
  @spec stop_reason(String.t() | nil) :: String.t()
  def stop_reason("stop"), do: "end_turn"
  def stop_reason("length"), do: "max_tokens"
  def stop_reason("tool_calls"), do: "tool_use"
  def stop_reason("content_filter"), do: "refusal"
  def stop_reason(nil), do: "end_turn"
  def stop_reason(other), do: other

  @doc "Map OpenAI usage → Anthropic usage (the fields Anthropic clients require)."
  @spec usage(map() | nil) :: map()
  def usage(nil), do: %{"input_tokens" => 0, "output_tokens" => 0}

  def usage(u) do
    %{
      "input_tokens" => u["prompt_tokens"] || u[:prompt_tokens] || 0,
      "output_tokens" => u["completion_tokens"] || u[:completion_tokens] || 0
    }
  end

  # === streaming: normalized chunks → Anthropic SSE events ===

  @doc """
  Build the Anthropic SSE preamble (message_start only). Content blocks are
  opened lazily by `ExLiteLLM.Anthropic.Stream` so tool-only / reasoning-only
  replies are not wrapped in an empty text block.
  """
  @spec stream_preamble(String.t()) :: [String.t()]
  def stream_preamble(model) do
    {state, frames} = ExLiteLLM.Anthropic.Stream.new(model) |> ExLiteLLM.Anthropic.Stream.preamble()
    _ = state
    frames
  end

  @doc "Translate one normalized text delta into an Anthropic content_block_delta event."
  @spec stream_text_delta(String.t()) :: String.t()
  def stream_text_delta(text), do: stream_text_delta(0, text)

  @spec stream_text_delta(non_neg_integer(), String.t()) :: String.t()
  def stream_text_delta(index, text) do
    sse("content_block_delta", %{
      "type" => "content_block_delta",
      "index" => index,
      "delta" => %{"type" => "text_delta", "text" => text}
    })
  end

  @spec stream_thinking_delta(non_neg_integer(), String.t()) :: String.t()
  def stream_thinking_delta(index, text) do
    sse("content_block_delta", %{
      "type" => "content_block_delta",
      "index" => index,
      "delta" => %{"type" => "thinking_delta", "thinking" => text}
    })
  end

  @spec stream_input_json_delta(non_neg_integer(), String.t()) :: String.t()
  def stream_input_json_delta(index, partial_json) do
    sse("content_block_delta", %{
      "type" => "content_block_delta",
      "index" => index,
      "delta" => %{"type" => "input_json_delta", "partial_json" => partial_json}
    })
  end

  @doc "Build the closing events (content_block_stop, message_delta, message_stop)."
  @spec stream_closing(String.t() | nil, map() | nil) :: [String.t()]
  def stream_closing(finish_reason, usage_map) do
    [
      sse("content_block_stop", %{"type" => "content_block_stop", "index" => 0})
      | stream_closing_after_blocks(finish_reason, usage_map)
    ]
  end

  @doc "message_delta + message_stop after content blocks have already been closed."
  @spec stream_closing_after_blocks(String.t() | nil, map() | nil) :: [String.t()]
  def stream_closing_after_blocks(finish_reason, usage_map) do
    [
      sse("message_delta", %{
        "type" => "message_delta",
        "delta" => %{"stop_reason" => stop_reason(finish_reason), "stop_sequence" => nil},
        "usage" => %{"output_tokens" => output_tokens(usage_map)}
      }),
      sse("message_stop", %{"type" => "message_stop"})
    ]
  end

  @doc "Anthropic error SSE event (used when the upstream 4xx after the stream started)."
  @spec stream_error(String.t(), String.t()) :: String.t()
  def stream_error(type, message) do
    sse("error", %{
      "type" => "error",
      "error" => %{"type" => type, "message" => message}
    })
  end

  defp output_tokens(nil), do: 0
  defp output_tokens(u), do: u["completion_tokens"] || u[:completion_tokens] || 0

  @doc "Format one Anthropic SSE frame (`event:` + `data:` lines)."
  @spec sse(String.t(), map()) :: String.t()
  def sse(event, data), do: "event: #{event}\ndata: #{Jason.encode!(data)}\n\n"

  # --- helpers ---

  defp choice_field(choice, key), do: choice[key] || choice[to_string(key)]

  defp rand, do: 12 |> :crypto.strong_rand_bytes() |> Base.url_encode64(padding: false)
end
